### ISSUE TITLE
Hooks to change Petting, Chat Bubble, Party Hat, and Emote Bubble positions for NPCs.

### ISSUE THREAD

**Issue by Rijam** [bluemagic123/tModLoader#4083]
### Do you intend to personally contribute/program this feature?

Yes

### I would like to see this change made to improve my experience with

Mod capability as a Modder

### Description

The ability to easily change the position and `isPetSmall` for petting Town Pets/Slimes.
The ability to easily change the position and SpriteEffect for the chat bubble, party hat, and emote bubble for NPCs.

*The NPCLoader code shown in the drop downs are just pseudocode. I know the NPC argument probably won't work as is.*

#### Petting

* `ModNPC.SetPettingInfo(ref int playerDistance, ref bool isPetSmall)`
* `GlobalNPC.SetPettingInfo(int npcType, ref int playerDistance, ref bool isPetSmall)`

Hooks to change how far/close the player stands to the Town Pet. The `isPetSmall` bool will make the player's arm either angled up or down.

<details><summary>Vanilla code Player.GetPettingInfo() **Click to Expand**</summary>

```diff
private void GetPettingInfo(int animalNpcIndex, out int targetDirection, out Vector2 playerPositionWhenPetting, out bool isPetSmall)
{
	NPC nPC = Main.npc[animalNpcIndex];
	targetDirection = ((nPC.Center.X > base.Center.X) ? 1 : (-1));
	int num = 36;
	isPetSmall = false;
	switch (nPC.type) {
		case 637: // <- Town Cat
			isPetSmall = true;
			num = 28;
			break;
		case 656: // <- Town Bunny
			isPetSmall = true;
			num = 24;
			break;
		case 670: // <- Nerdy Slime
		case 678: // <- other Town Slimes
		case 679:
		case 680:
		case 681:
		case 683:
			isPetSmall = true;
			num = 26;
			break;
		case 682: // <- Surly Slime
			isPetSmall = true;
			num = 22;
			break;
		case 684: // <- Squire Slime
			isPetSmall = true;
			num = 20;
			break;
	}
	
+	NPCLoader.SetPettingInfo(nPC, ref num, ref isPetSmall);

	playerPositionWhenPetting = nPC.Bottom + new Vector2(-targetDirection * num, 0f);
}
```
	
</details>
  
#### Chat Bubble

* `ModNPC.NPCChatBubblePosition(ref Vector2 position, ref SpriteEffects spriteEffects)`
* `GlobalNPC.NPCChatBubblePosition(int npcType, ref Vector2 position, ref SpriteEffects spriteEffects)`
	
This is the chat bubble that shows up when you hover over the Town NPC. The hooks lets you change the final position before the bubble is drawn. The SpriteEffects give you the power to flip it horizontally if you wish.

<details><summary>Vanilla code Main.DrawNPCChatBubble() **Click to Expand**</summary>
  
```diff
private static void DrawNPCChatBubble(int i)
{
	int num = -(npc[i].width / 2 + 8);
	float num2 = npc[i].position.Y - (float)TextureAssets.Chat.Height() - (float)(int)screenPosition.Y;
	if (npc[i].type == 637 && npc[i].ai[0] == 5f) // <- Town Cat
		num2 -= 18f;

	SpriteEffects spriteEffects = SpriteEffects.None;
	if (npc[i].spriteDirection == -1) {
		spriteEffects = SpriteEffects.FlipHorizontally;
		num = npc[i].width / 2 + 8;
	}

	if (player[myPlayer].gravDir != 1f) {
		spriteEffects |= SpriteEffects.FlipVertically;
		num2 = (float)screenHeight - num2 - (float)TextureAssets.Chat.Height();
	}

	Vector2 position = new Vector2(npc[i].position.X + (float)(npc[i].width / 2) - screenPosition.X - (float)(TextureAssets.Chat.Width() / 2) - (float)num, num2);
	
+	NPCLoader.NPCChatBubblePosition(npc[i], ref position, ref spriteEffects);
	
	spriteBatch.End();
	spriteBatch.Begin(SpriteSortMode.Deferred, null, SamplerState.PointClamp, null, null, null, GameViewMatrix.ZoomMatrix);
	PlayerInput.SetZoom_UI();
	spriteBatch.Draw(TextureAssets.Chat.Value, position, new Microsoft.Xna.Framework.Rectangle(0, 0, TextureAssets.Chat.Width(), TextureAssets.Chat.Height()), new Microsoft.Xna.Framework.Color(mouseTextColor, mouseTextColor, mouseTextColor, mouseTextColor), 0f, default(Vector2), 1f, spriteEffects, 0f);
	spriteBatch.End();
	spriteBatch.Begin(SpriteSortMode.Deferred, null, null, null, null, null, _uiScaleMatrix);
}
```

</details>

#### Party Hat	

* `ModNPC.NPCPartyHatPosition(ref Vector2 position, ref SpriteEffects spriteEffects)`
* `GlobalNPC.NPCPartyHatPosition(int npcType, ref Vector2 position, ref SpriteEffects spriteEffects)`

More info mentioned below. The hooks lets you change the final position before the party hat is drawn. The SpriteEffects give you the power to flip it horizontally if you wish. 
    
<details><summary>Vanilla code Main.DrawNPCExtras() **Click to Expand**</summary>
  
```diff
protected void DrawNPCExtras(NPC n, bool beforeDraw, float addHeight, float addY, Microsoft.Xna.Framework.Color npcColor, Vector2 halfSize, SpriteEffects npcSpriteEffect, Vector2 screenPosition)
{
	if (!beforeDraw && n.UsesPartyHat()) {
		int frame = n.frame.Y / n.frame.Height; // renamed: num
		int[] array = NPCID.Sets.TownNPCsFramingGroups[NPCID.Sets.NPCFramingGroup[n.type]];
		if (frame >= array.Length)
			frame = 0;

		Texture2D value = TextureAssets.Extra[72].Value;
		int num2 = 0;
		switch (n.GetPartyHatColor()) {
			// party hat colors
		}
		
		Microsoft.Xna.Framework.Rectangle rectangle = value.Frame(20, 1, num2 % 20);
		rectangle.Width -= 2;
		rectangle.Height -= 2;
		int xOffset = 0; // renamed: num3
		int yOffset = 0; // renamed: num4
		int spriteDirection = n.spriteDirection; // renamed: num5
		if (n.type == 663) // <- Princess
			xOffset = 1;

		if (n.type == 637) { // <- Town Cat
			xOffset = 6;
			switch (frame) {
				case 19:
				case 22:
				case 23:
				case 24:
				case 25:
				case 26:
				case 27:
					xOffset -= 2;
					break;
				case 11:
				case 12:
				case 13:
				case 14:
				case 15:
					xOffset += 2;
					break;
			}
		}

		if (n.type == 638) // <- Town Dog
			xOffset = 12;

		if (n.type == 656) { // <- Town Bunny
			xOffset = 6;
			switch (frame) {
				case 1:
				case 2:
				case 3:
					xOffset -= 2;
					break;
				case 18:
				case 19:
				case 20:
				case 21:
				case 22:
				case 23:
				case 24:
				case 25:
					xOffset -= 4;
					break;
				case 8:
					xOffset -= 2;
					break;
			}
		}

		if (NPCID.Sets.IsTownSlime[n.type]) { // <- Town Slimes
			spriteDirection *= -1;
			switch (n.type) {
				default:
					xOffset = 1;
					break;
				case 670:
					xOffset = -1;
					break;
				case 681:
					xOffset = -2;
					break;
				case 682:
					xOffset = -1;
					yOffset = 2;
					break;
				case 683:
					xOffset = -1;
					yOffset = -2;
					break;
				case 684:
					xOffset = -4;
					break;
				case 679:
					break;
			}
		}

		if (n.IsShimmerVariant) {
			switch (n.type) {
				case 38: // <- Demolitionist
					xOffset += 2;
					yOffset = -2;
					break;
				case 178: // <- Steampunker
				case 228: // <- Witch Doctor
					xOffset = -4;
					yOffset = -4;
					break;
				case 107: // <- Goblin Tinkerer
					xOffset = 2;
					yOffset = -6;
					break;
				case 54: // <- Clothier
					yOffset = -6;
					break;
				case 160: // <- Truffle
					yOffset = -2;
					break;
				case 108: // <- Wizard
				case 124: // <- Mechanic
				case 208: // <- Party Girl
				case 209: // <- Cyborg
				case 227: // <- Painter
					yOffset = -4;
					break;
			}
		}
		Vector2 vector = n.Top + new Vector2(-2 * spriteDirection, n.gfxOffY);
		vector.X += xOffset * spriteDirection;
		vector.Y += array[frame];
		vector.Y += yOffset;
		vector.Y += NPCID.Sets.HatOffsetY[n.type]; // The set that we can currently control.
		int yOffsetFromSitting = 0; // renamed: num6
		if (n.ai[0] == 5f) { // <- Town NPC is sitting
			yOffsetFromSitting = -4;
			if (n.type == 38) // <- Demolitionist
				yOffsetFromSitting = -8;

			if (n.type == 124) // <- Mechanic
				yOffsetFromSitting = -2;

			if (n.type == 550) // <- Tavernkeep
				yOffsetFromSitting = -4;

			if (n.type == 588) // <- Golfer
				yOffsetFromSitting = -4;

			if (n.type == 108 || n.type == 178) // <- Wizard || Steampunker
				yOffsetFromSitting = -6;

			if (n.type == 637) // <- Town Cat
				yOffsetFromSitting = -12;

			if (n.type == 663) // <- Princess
				yOffsetFromSitting = -8;
		}

		vector.Y += yOffsetFromSitting;
		if (n.type == 229 && n.ai[0] == 12f) // Pirate, attacking related
			vector.X -= spriteDirection * 4;

		if (n.type == 550 && n.ai[0] == 5f) // Tavernkeep, sitting
			vector.X += spriteDirection * 7;

		Vector2 origin = rectangle.Size() - new Vector2(rectangle.Width / 2, 12f);
		int xOffsetAgain = 0; // renamed: num7
		switch (n.type) {
			case 550: // <- Tavernkeep
				xOffsetAgain = -4;
				break;
			case 588: // <- Golfer
				xOffsetAgain = 0;
				break;
			case 227: // <- Painter
				xOffsetAgain = -4;
				break;
			case 228: // <- Witch Doctor
				xOffsetAgain = -2;
				break;
			case 17: // <- Merchant
			case 18:
			case 19:
			case 20:
			case 22:
			case 124:
			case 229:
			case 353:
			case 633:
			case 637:
			case 638:
			case 656:
			case 670:
			case 678:
			case 679:
			case 680:
			case 681:
			case 682:
			case 683:
			case 684:
				xOffsetAgain = -1;
				break;
			case 37:
			case 38:
			case 54:
			case 107:
			case 108:
			case 160:
			case 207:
			case 209:
				xOffsetAgain = -3;
				break;
			case 178:
			case 208:
			case 369:
				xOffsetAgain = 1;
				break;
		}

		vector.X += xOffsetAgain * spriteDirection;
		vector.X += 4 * spriteDirection;
		SpriteEffects spriteEffects = npcSpriteEffect;
		if (NPCID.Sets.IsTownSlime[n.type])
			spriteEffects ^= SpriteEffects.FlipHorizontally;

		float num8 = 1f;
		if (n.shimmerTransparency > 0f)
			num8 *= 1f - n.shimmerTransparency;

+		NPCLoader(npc, ref vector, ref spriteEffects);

		spriteBatch.Draw(value, new Vector2(vector.X - screenPosition.X, vector.Y - screenPosition.Y), rectangle, npcColor * n.Opacity * num8, 0f, origin, n.scale, spriteEffects, 0f);
	}

	//... Diva slime special drawing
	
	//... Attacking drawing stuff
}
```
	
</details>
  
##### Option 1
    
* `ModNPC.NPCEmoteBubblePosition(ref SpriteEffects effect, ref float xMultiplier, ref float yOffset)`
* `GlobalNPC.NPCEmoteBubblePosition(int npcType, ref SpriteEffects effect, ref float xMultiplier, ref float yOffset)`

Vanilla multiplies the X position by 0.75f and the Y position is always 2f. The hooks lets you change the X multiplier and the Y position before the bubble is drawn. The SpriteEffects give you the power to flip it horizontally if you wish.
    
<details><summary>Vanilla code GameContent.UI.EmoteBubble.GetPosition() **Click to Expand**</summary>
  
```diff
private Vector2 GetPosition(out SpriteEffects effect)
{
	switch (anchor.type) {
		case WorldUIAnchor.AnchorType.Entity:
			effect = ((anchor.entity.direction != -1) ? SpriteEffects.FlipHorizontally : SpriteEffects.None);
+			float xMultiplier = 0.75f;
+			float yOffset = 2f;			
+			NPCLoader.NPCEmoteBubblePosition(npc, ref effect, ref xMultiplier, ref yOffset);			
-			return new Vector2(anchor.entity.Top.X, anchor.entity.VisualPosition.Y) + new Vector2((float)(-anchor.entity.direction * anchor.entity.width) * 0.75f, 2f) - Main.screenPosition;
+			return new Vector2(anchor.entity.Top.X, anchor.entity.VisualPosition.Y) + new Vector2((float)(-anchor.entity.direction * anchor.entity.width) * xMultiplier, yOffset) - Main.screenPosition;
		case WorldUIAnchor.AnchorType.Pos:
			effect = SpriteEffects.None;
			return anchor.pos - Main.screenPosition;
		case WorldUIAnchor.AnchorType.Tile:
			effect = SpriteEffects.None;
			return anchor.pos - Main.screenPosition + new Vector2(0f, (0f - anchor.size.Y) / 2f);
		default:
			effect = SpriteEffects.None;
			return new Vector2(Main.screenWidth, Main.screenHeight) / 2f;
	}
}
```

</details>
  
##### Option 2 (I like this one better)

* `ModNPC.NPCEmoteBubblePosition(ref Vector2 position, ref SpriteEffects effect)`
* `GlobalNPC.NPCEmoteBubblePosition(int npcType, ref Vector2 position, ref SpriteEffects effect)`
	
More consistent with the other hooks. The hooks lets you change the position before the bubble is drawn. The SpriteEffects give you the power to flip it horizontally if you wish.

<details><summary>Vanilla code GameContent.UI.EmoteBubble.Draw() **Click to Expand**</summary>
  
```diff
private void Draw(SpriteBatch sb)
{
	Texture2D value = TextureAssets.Extra[48].Value;
	SpriteEffects effect = SpriteEffects.None;
	Vector2 position = GetPosition(out effect);
	position = position.Floor();
	
+	NPCLoader.NPCEmoteBubblePosition(npc, ref position, ref effect); // Could place it before the position = position.Floor() line instead.
	
	bool flag = lifeTime < 6 || lifeTimeStart - lifeTime < 6;
	Rectangle value2 = value.Frame(8, EMOTE_SHEET_VERTICAL_FRAMES, (!flag) ? 1 : 0);
	Vector2 origin = new Vector2(value2.Width / 2, value2.Height);
	if (Main.player[Main.myPlayer].gravDir == -1f) {
		origin.Y = 0f;
		effect |= SpriteEffects.FlipVertically;
		position = Main.ReverseGravitySupport(position);
	}
	
	// ...
}
```

</details>

### What does this proposal attempt to solve or improve?

Currently, changing the petting settings and emote bubble position require a detour. Changing chat bubble and party hat position requires an IL edit (unless you copy paste the entire thing into a detour or something).

The current only ways to edit the position of the party hat include `NPCID.Sets.HatOffsetY[]` and `NPCID.Sets.NPCFramingGroup[]`. The first one is only a Y offset and the second one only has 8 options available. Since they are sets, you are encuraged to put them in SetStaticDefaults which means they only run once on mod load. [More info about NPCFramingGroup in my Town NPC Guide](https://gist.github.com/Rijam/c67eb86eb3aef9db7ac474ad5ef7dfba#party-hat)

If the design for your Town NPC/Pet/Slime doesn't match a vanilla Town NPC/Pet/Slime, such as having a head that isn't centered horizontally or has a different walk animation, there isn't much you can do to customize these extra things. 

Some specific examples:
I've made a clone of the Town Cat for Town NPC guide. The raised position when the cat is sitting on the chair is achieved with a detour of `Main.NPCAddHeight()`, but as you can see the chat bubble isn't raised up either. You can also see that the party hat isn't on the cat clone's head because its head isn't centered horizontally.
![CatAndCatClone](https://github.com/tModLoader/tModLoader/assets/11262234/68aeaebe-95e7-4849-9e45-4d902f8528a9)
Another example is the Draken from SGAmod (1.3 mod, isn't available for 1.4). His head is not centered so the best we can do is put the party hat on his back. The emote bubbles also come out of his behind instead of his head.
![DrakenExample](https://github.com/tModLoader/tModLoader/assets/11262234/cf64658f-1538-4e2f-8ab9-7b40ff52e313)

### Which (other) solutions should be considered?

Of course the names of the hooks and arguments could be changed.

These hooks only let you change the position (and SpriteEffects) of the thing being drawn, but not the scale, rotation, origin, etc. The petting hook could be the position instead of just the distance.

A hook for modifying `Main.NPCAddHeight()` could also be added because there are Town NPC/Pet/Slime related things in it.

**Comment by Chicken-Bones** [comment]
The pet's height and whether it's small should be type based, perhaps an `NPCID.Sets` entry.
The rest seem well motivated. I'm fine with `(ref Vector2 position, ref SpriteEffects flip)` where `position` is in world space. I'd love for this to be an offset, but there's no consistent place in an NPC that's reasonable to offset from.

I'm slowly coming around to more and more of these 'overly specific hooks' where they're needed

**Comment by Rijam** [comment]
Yeah, these hooks are pretty niche and most people probably won't have to use them.

For the petting, are you thinking more of a `int[] PlayerDistanceWhilePetting = Factory.CreateIntSet(36, ...)` and a `bool[] IsPetSmallForPetting = Factory.CreateBoolSet(...)`? That works for me. The ability to change the values during run time is not important for 99% of people. They can still detour the vanilla method if they really need that.

I was trying to go with the minimum patches as possible so editing the final position was the easiest way. But I agree that the three positions aren't consistent, so the offsets would kind of be arbitrary. But, we could still do it just like:
```diff
	vanilla code here...
+	Vector2 offset = Vector2.Zero;
+	NPCLoader.WhateverHookVoid(npc, ref offset, ref effects);
+	position += offset;
	spriteBatch.Draw(...)
```
Or make the hooks return Vector2s
```diff
	vanilla code here...
+	position += NPCLoader.WhateverHookReturnsVector2(npc, ref effects);
	spriteBatch.Draw(...)
```

I've also edited the original post to have two specific examples.

**Comment by JavidPack** [comment]
Fixed in #4222
